nodejs/src/stream_manager.rs

```diff
@@ -450,23 +410,47 @@ impl StreamManager {
         }
     }
 
-    pub fn js_list_scopes(mut cx: FunctionContext) -> JsResult<JsArray> {
-        use std::convert::TryInto;
-
+    pub fn js_list_scopes(mut cx: FunctionContext) -> JsResult<JsPromise> {
         let stream_manager = cx.this().downcast_or_throw::<JsBox<StreamManager>, _>(&mut cx)?;
 
-        let scopes = stream_manager.list_scopes();
-        let scopes_length: u32 = match scopes.len().try_into() {
-            Ok(val) => val,
-            Err(_err) => return cx.throw_error("Too many scopes"),
-        };
-
-        let array: Handle<JsArray> = JsArray::new(&mut cx, scopes_length);
-        for (pos, e) in scopes.iter().enumerate() {
-            let scope_name = cx.string(e.name.clone());
-            array.set(&mut cx, pos as u32, scope_name)?;
-        }
-        Ok(array)
+        let channel = cx.channel();
+        let (deferred, promise) = cx.promise();
+
+        // spawn an `async` task on the tokio runtime.
+        use std::sync::Arc;
+        let cf = Arc::new(stream_manager.cf.to_async());
+        let handle = stream_manager.cf.runtime_handle();
+        // enter the runtime context and call spawn with timeout
+        let _rt = handle.enter();
+        tokio::spawn(time::timeout(
+            Duration::from_millis(TIMEOUT_IN_MILLISECONDS),
```

Comment:
I don't think it's our job to put a timeout here. If the user wants to limit their blocking they can do that in their own code. Especially as 1s is just totally arbitrary.

Reply:
If users want to limit they can only do it in nodejs. And in nodejs, you just neglect the result of the promise after some timeout. But by doing this way, the internal connection is still open and may or may not be timeout after some uncontrolled time periods.

With `Duration::from_millis(TIMEOUT_IN_MILLISECONDS)`, it is certain that there will be no more internal connection left unmanaged even if the user doesn't care about it anymore. And this timeout is similar to the python wrapper `const TIMEOUT_IN_SECONDS: u64 = 120;` where the limit is hardcoded.

But I agree we could use other options to control the actual timeout instead of a hardcoded 1 second.

Reply:
It should work as a sanity check. I updated the time to 120 sec.